Ship tags and descriptions

Latentish opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## About The Pull Request
This adds descriptions and tags to every ship in the game that is not a subshuttle. It also makes these viewable in the Ship Purchase, Ship Join, and Shuttle Manipulator tabs. It also reworks the shuttle manipulator template tab, and adds the ability to add or remove tags from ships manually in the ship editor.
![6dd8e53c4b8224837f1d9748c4a37ccf](https://github.com/shiptest-ss13/Shiptest/assets/55075747/f1890609-6140-4a1e-bcb0-9827284ff676)
![77def6942b2f4c4fa4424f57b80f9058](https://github.com/shiptest-ss13/Shiptest/assets/55075747/20cfdbc3-2089-44fa-b4c0-d27511daa076)
![cc78d07d7c2e27526ab4c3756dcb7831](https://github.com/shiptest-ss13/Shiptest/assets/55075747/097a1a31-ba8e-44bd-9cdd-6a4565cc4eea)
![eab030d58d6187c8111162a141f761f5](https://github.com/shiptest-ss13/Shiptest/assets/55075747/fa76f828-a701-4799-9af6-17de36f08934)

Search features and other ship purchasing tgui related changes will be added in separate PR.
<!-- Describe The Pull Request. Please be sure every change is documented or this can delay review and even discourage maintainers from merging your PR! -->

## Why It's Good For The Game
This allows people to get a better sense of what they're buying. It also reworks some years old shuttle manipulator ui.
<!-- Please add a short description of why you think these changes would benefit the game. If you can't justify it in words, it might not be worth adding. -->

## Changelog

:cl:phoaly
add: Added descriptions to all jsons
add: Added tags to all ship jsons
add: Added Ship Tags and descriptions to the ship purchase/join/shuttle manipulator menus
add: Added tag manipulation to the ship editor
tweak: Tweaks the UI of the shuttle manipulator
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

Latentish committed:
Adds the base

-Trying to add shiptags

Latentish committed:
adds delimiter

-Fixes an error

Latentish committed:
Another formatting oopsie

-Removes trailing }

Latentish committed:
Trailing commo for the linter gods

-Kills a trailing comma

Latentish committed:
Fixes case, and adds metis

- in ShipSelect.js changes the Label to label for ship tags.
- adds tags and a description for the metis

Latentish committed:
Tries to enable tag array

- changes ship tag config type to "array"
-Changes how these are handled in ship_select.js

Latentish committed:
Revert "Tries to enable tag array"

This reverts commit bafbdab1e734c00b74b280cd22cff4a2bf9debbb.

Latentish committed:
Revert "Revert "Tries to enable tag array""

This reverts commit 51653e1580c91934a36c8599cf823730861fb877.

Latentish committed:
update the array called

-titular

Latentish committed:
Makes them show the fuck up

-TAGS CAN BE SEEEEEEN

Latentish committed:
Adds descriptions and tags to all jsons

Titular

Latentish committed:
Switches an array

-Swaps template.tags for joining a ship to selectedShip.tags
-This makes the join button work again

Latentish committed:
Join ship data

-Makes descriptions and tags show up for joinable ships
-Does this by adding them to the join ship data, referencing the template material created for purchasing.

Latentish committed:
Merge remote-tracking branch 'upstream/master' into Ship-Tags-and-Descriptions

Latentish committed:
Merge remote-tracking branch 'upstream/master' into Ship-Tags-and-Descriptions

Latentish committed:
Merge branch 'master' into Ship-Tags-and-Descriptions

Signed-off-by: Latency <55075747+Latentish@users.noreply.github.com>

Latentish committed:
Last round of touchups

-Adds tags and a description for the Beluga
- Moves ship memos above the description
- Adds a txt document describing how to tag and describe ships. (mostly how to tag them)

tmtmtl30 commented on tgui/packages/tgui/interfaces/ShipSelect.js:
```diff
@@ -101,6 +101,10 @@ export const ShipSelect = (props, context) => {
+                  {(selectedShip.tags && selectedShip.tags.join(', ')) ||
+                    'No Tags Set'}
+                </LabeledList.Item>
                 <LabeledList.Item label="Ship Memo">
```
you mentioned wanting to move the memo up, right?

tmtmtl30 commented on code/datums/shuttles.dm:
```diff
@@ -194,6 +195,7 @@
 	.["templateName"] = name
 	.["templateShortName"] = short_name
 	.["templateDescription"] = description
+	.["templateTags"] = tags
```
without modifying the template modification js (ShipEditor.js, i think), this won't do anything

Latentish replied:
Ship editor is fully supported now.

tmtmtl30 commented on code/datums/shuttles.dm:
```diff
@@ -227,6 +229,10 @@
 			description = params["new_template_description"]
 			update_static_data(usr, ui)
 			return TRUE
+		if("setTemplateTags")
```
same here -- you'd need to make a button that fires off an act() containing, as a list, the new tags for the template

Latentish replied:
Same, it's fully supported now, and this has been split into addTemplateTags and removeTemplateTags.

Latentish committed:
Tries to add functionality to the editor and manipulator

-Adds a "set tags" to the shipeditor
-Fails to add a "ship tags" to ship templates in the shuttle manipulator.

Latentish committed:
Some patchie for the soul

-Makes tags a var/list in shuttles.dm
-Adds S.Tags to the subsystem shuttle.dm
-Adds formatting for tags in the shuttle manipulator
-Tries and fails to add nullchecking.

Latentish commented:
![image](https://github.com/shiptest-ss13/Shiptest/assets/55075747/613d87fd-759f-4f23-8bea-5be8251660c4)
I'll try to fix this tomorrow.

Latentish committed:
A lot

- Removes "admin_notes", it wasn't used anywhere, or implemented to any extent.
- Makes the shuttle manipulator templates a series of collapsibles full of details rather than Sections.
- Adds functionality to change descriptions of ships, add tags, and remove tags to the ship editor

Latentish committed:
Merge remote-tracking branch 'upstream/master' into Ship-Tags-and-Descriptions

Latentish commented:
![image](https://github.com/shiptest-ss13/Shiptest/assets/55075747/d1e5f934-f610-4ef0-a624-8cf90fcf010a)
This is the only thing I personally touched from the upstream merge. Specifically changing "Info" to "Details".

Latentish committed:
Stop yelling at me!!!!

-WHAT DO YOU MEAN PRETTIER IS UPSET

Latentish committed:
I hate prettier I hate prettier I hate prettier

-FIRMLY GRASPIT

Latentish commented:
I intend to add extra features in another pr once this is merged.
Tags being generated as buttons is a top priority for that pr.
Implementing a tag toggle system in the ship editor is pretty low priority (as tag editing is extremely marginal in usecase anyway).
Adding a tag search system in the three places it'd be useful for (shuttle manipulator, ship purchase, and ship join) (High priority).
Adding a link to the webmaps or embedding images of the ships as previews during ship purchase. (on the fence about this)
Adding a way to set up a ship's join settings and memo on purchase (this is gonna be difficult to figure out).

But for now I want to get this merged because these features are tangential to this pr. The sooner this is merged, the less technical debt accrues for ship descriptions and tags needing to be set for new ships. (I'd prefer codeowners set their own descriptions and tags instead of myself.)

tmtmtl30 commented:
LGTM
